Welcome to Machine Learning for Saja (ML4S) crash course! few things to know before we get started:

*   This course will focus on the practical side of Machine Learning (you will be useful basically)
*   The target variable of this course is a very smart woman. Hence, if something feels not very straightforward.. read it again.
* The practice problems will be added at the end of the notebook and are usually similar to the problem disccused. However, twists may appear at times.

without any further delay lets do some data science!


Introduction:

*   Machine Learning is a subset of data science and it is the science of letting the machine figure out how a task is done. <br> In simple terms if we want the answer for 5 + 3 we would typically do the following <br> `a = 5, b = 3` <br> `print(a+b)` <br> and we get our 8. <br><br> In machine learning however, we do not tell the machine what to do. We let the machine figure out what to do. <br>To illustrate, we feed the following equation: <br> `a ? b = 8` <br> we want the machine to figure out what to do between a and b to get this 8 <br>we call variables [a, b, and 8] our ***predictors***. Meanwhile, we call ? our ***target variable*** that we want to obtain.








There are three types of machine learning Algorithms. An algorithm is considered ***supervised*** when you have a target variable. <br> Example: <br>

`inputs = [smart, pretty, evil, ugly]` <br>
`labels = [1, 1, 0, 0]` <br>

Here each input is mapped to a label marking whether its a positive or a negative word. Since labels exist, this problem is supervised. <br>

When the labels are not available, we consider the algorithm ***unsupervised***. Unsupervised problems are solved through ***clustering***. In other words, grouping inputs that look similar together. <br> Example: <br>

`inputs = [-2, -4, 7, 6, 5]` <br>

This would result two clusters where cluster1 = [-2,-4] and cluster2 = [7,6,5]. A cluster can then be given a label; e.g. Negative_nums = cluster1. <br>

A third and final type is semi-supervised algorithms. It is applied when the data is a mixture of both the previous types. It is usually applied on two stages where labels are used first to train a model before using the same model to inference the unlabled data and retrain. <br> Example: <br>


`Age = [1993, 1996, 2001, 2008, 1999, 1970, ...,1994, 1975]` <br>
`labels = [Y, null, Z, Z, null, X,..., Y, null]` <br>

Seperate labelled from unlabelled data: <br>

`Age_labelled = [1993, 2001, 2008, 1970, ..., 1994]` <br>
`labels_1 = [Y, Z, Z, X,..., Y]` <br>

`Age_unlabelled = [1996, 1999,..., 1975]` <br>

Train a model using the labelled data <br>
`model.train(Age_labelled, labels_1)` <br>
use the trained model to label the unlabelled years <br>
`new_labels = model.predict(Age_unlabelled)` <br>
add the new labels to the dataset and train again now using the full dataset. <br>
`Age = Age_labelled + new_labels` <br>
`labels = labels_1 + new_labels` <br>
`model.train(Age, labels)` <br>

Our main focus will be on supervised learning so don't worry if the others are not clear. <br> This is just some theory that you should know. Now lets start!

TOPIC 000: Introduction to Python, Numpy, and Pandas

Python is a very straight forward language. Unlike other languages like the rusty C++ or your beloved Java, this language is very close to your everyday dialog. In addition, very resourceful. This is important because if you do not know how to do something just google it and you will get your answer.

fun fact: A great man once said "if you memorize every function by heart.. you are not doing it right" - Diaa

In [ ]:
#in python, no need to declare the varaible type.
a = 5
b = "Ahmed"
c = True
d = 43.1
print(a, b, c, d)

5 Ahmed True 43.1


Colab notebook is a cell based IDE. This means that you can execute codes by cell (order of cells location does not matter). Just hover your mouse over the cell and click on the run button or press Ctrl + Enter to run a cell.

One more thing you need to know about this language is that indentation is very important. (See below)

In [ ]:
if a == 5:
#print(1) ##this will generate indentation error
  print(1) #you cannot return outside a function in python as well

def func(a): #you can specify the datatype here but not needed. it is python!
  return a #this is fine

func(5)

1


5

Now lets focus on more useful things in ML. Python is something that you will learn as we go

One of the most common libraries in Data Science is NumPy (Num Pi not Num Pee).
NumPy is very good because it can do tasks that would take 15 lines of code in 1 line. An example is shown below.

In [ ]:
import numpy as np #in python you need to import what you want to use. Here I am importing numpy but I am too lazy to call it each time so I am naming it np instead
#lets create a 3x2 Matrix of zeros perhaps?

zeros = np.zeros([3,2])

zeros #typing a variable name without print will print that variable (works for one var only, having many this way will print the last one only)

array([[0., 0.],
       [0., 0.],
       [0., 0.]])

Numpy has some important functions that you should be aware of. One of the most important ones is np.shape which returns the shape of your array.

In [ ]:
zeros.shape

(3, 2)

numpy allows you to reshape as well to wtv as long as it is possible

In [ ]:
zeros = zeros.reshape([2,3])
zeros.shape

(2, 3)

Practice Problem 1)

*   Below is a random list. you are asked to do the following:
1.   Convert the list to a numpy array
2.   Sort the array and save it in another variable name
3.   perform element-wise addition
4.   find the mean of the summed array
5.   find the maximum of the summed array
6.   Print the list and the new variables



* (make use of numpy and remember the wise man words)
* every task takes one line only

In [ ]:
A = [3,5,9,2,4,0] #Expected output is shown below

[3, 5, 9, 2, 4, 0] [3 5 9 2 4 0] [0 2 3 4 5 9] [ 3  7 12  6  9  9] 7.666666666666667 12


In [ ]:
#import numpy as np
A = [3,5,9,2,4,0]
array = np.array(A)
sort = np.sort(A)
added = sort+sort #oh so sort+array
mean = np.mean(added)
max = np.max(added)

print(A)
print(array)
print(sort)
print(added)
print(mean)
print(max)

# if Saja.cheated:
#   print("Bombastic Side Eye (Do not cheat.. go and try yalla!)")
# else:
#   print("Well done Saja!, see numpy can be very useful as long as you search!")

A = [3,5,9,2,4,0]
B = np.array(A)
C = np.sort(B)
D = C + B
mean = np.mean(D)
max = np.max(D)

Now I will move to a much more important library. Mastering this library is mastering data science. Welcome to Pandas!

*   Everything that you can do in Numpy can be done in Pandas
*   Think of Pandas as the excel of python
*   it is very useful because this is how will you read data from the database for example.

to start, download the following dataset and lets do some work:

https://www.kaggle.com/datasets/aliibrahim10/anime-ratings



Once you download it and unzip it. press on the file button on the left panel (under {x}) then right click and upload the csv file.

In [ ]:
import pandas as pd #importing what we want
df = pd.read_csv("/content/MALratings.csv")
#df stands for dataframe (a sheet basically)
#read_csv is a function in pandas (works for any other format [xlsx etc])
#read_csv takes a path (you can obtain it by right clicking on the file and copy path (string))

In [ ]:
df

Title  \
0             Kaguya-sama wa Kokurasetai: Ultra Romantic   
1                             3-gatsu no Lion 2nd Season   
2      Gintama Movie 2: Kanketsu-hen - Yorozuya yo Ei...   
3                                               Gintama.   
4                                         Koe no Katachi   
...                                                  ...   
20338                                        Steins;Gate   
20339                                         High Score   
20340                                           Noraneko   
20341             Suki de Suki de, Suki de The Animation   
20342                            Koihime†Musou OVA Omake   

                               Genres   Rank  Popularity  Score Episodes  \
0                     Comedy,Suspense     13         404   8.99       12   
1                 Drama,Slice of Life     14         533   8.96       22   
2                Action,Comedy,Sci-Fi     19         939   8.93        1   
3                Action,Comedy,Sci-Fi     12         735   8.99       12   
4                                 NaN     15          19   8.96        1   
...                               ...    ...         ...    ...      ...   
20338           Drama,Sci-Fi,Suspense      4          13   9.08       24   
20339            Comedy,Slice of Life  10527        6095   5.58        8   
20340                             NaN   8706        5064   6.04        1   
20341                             NaN     /A        5329   6.81        2   
20342  Adventure,Comedy,Fantasy,Ecchi   8328       10052   6.12        1   

        Episode length Release Date  
0      23 min. per ep.  Spring 2022  
1      25 min. per ep.    Fall 2017  
2        1 hr. 50 min.          NaN  
3      24 min. per ep.  Winter 2017  
4        2 hr. 10 min.          NaN  
...                ...          ...  
20338  24 min. per ep.  Spring 2011  
20339   3 min. per ep.    Fall 2011  
20340          35 sec.          NaN  
20341  27 min. per ep.          NaN  
20342          12 min.          NaN  

[20343 rows x 8 columns]

now we have our dataset! and its large (20343 records!) joking this is small. usually we will deal with millions. (will do this to you next time hehe)

In [ ]:
df.head(6) #prints first 6 records. if left empty (df.head()) then default is 5. df.tail does the opposite obviously

Title                Genres  \
0         Kaguya-sama wa Kokurasetai: Ultra Romantic       Comedy,Suspense   
1                         3-gatsu no Lion 2nd Season   Drama,Slice of Life   
2  Gintama Movie 2: Kanketsu-hen - Yorozuya yo Ei...  Action,Comedy,Sci-Fi   
3                                           Gintama.  Action,Comedy,Sci-Fi   
4                                     Koe no Katachi                   NaN   

  Rank  Popularity  Score Episodes   Episode length Release Date  
0   13         404   8.99       12  23 min. per ep.  Spring 2022  
1   14         533   8.96       22  25 min. per ep.    Fall 2017  
2   19         939   8.93        1    1 hr. 50 min.          NaN  
3   12         735   8.99       12  24 min. per ep.  Winter 2017  
4   15          19   8.96        1    2 hr. 10 min.          NaN

How can we manipulate this data? One important concept in ML is variables type:

*   Variables can be numerical (numbers) or categorical (strings)
*   Creating models for numbers is easy but for strings???
*   This is where encoding comes where u map each string to some number.
*   We will cover this later. For now lets drop categorical data.



In [ ]:
df.dtypes

Title              object
Genres             object
Rank               object
Popularity          int64
Score             float64
Episodes           object
Episode length     object
Release Date       object
dtype: object

object is categorical. It makes sense that title and genres are strings but what about rank? hmmm smth is wrong yeah? its clearly a number.



*   this can happen if some entries in Rank column are null, empty, or NAN
*   if so we can use df['Rank'].fillna(0) and every null value will become a 0

lets have a look



In [ ]:
df['Rank'].isnull() #this returns an array that has False if the variable is null

0        False
1        False
2        False
3        False
4        False
         ...  
20338    False
20339    False
20340    False
20341    False
20342    False
Name: Rank, Length: 20343, dtype: bool

this is confusing since the dataset is big. lets make our life easier by first converting the column to a row using .values and then checking if any value is true using .any (i googled this)

In [ ]:
df['Rank'].isnull().values.any()

False

hmmm all good it seems.. then whats the problem???

*   okay lets convert the type. if there is an anomly we will get an error and figure it out



In [ ]:
df['Rank'] = pd.to_numeric(df['Rank'])

ValueError: ignored

Practice Problem 2.
Ahaaaa here we go! there is some "/A" string. Okay lets drop all rows that has this string.

In [ ]:
# df["Rank"].str.contains("/A") checks if we have the value "/A"
# we check if it is false(does not contan "/A" and we assign it back to df )
# now all the rows will only be the ones where "/A" does not exist

df = df[df["Rank"].str.contains("/A")== False]

In [ ]:
#@title Solution
df = df[df["Rank"].str.contains("/A")== False]

In [ ]:
#lets test now
df['Rank'] = pd.to_numeric(df['Rank'])

perfect! now do the following

*   Episodes seems to be numbers too? whats going on there? Fix it
*   Once done convert Episodes to numeric and check data types again
*   drop all categorical variables
*   drop all rows that contain nulls
*   Once done export the dataframe and download it.



000 Ends here. Only when I recieve the correct dataframe I will provide 001. Good lucksend

In [8]:
import pandas as pd
import os
df = pd.read_csv(r"/content/MALratings.csv")

In [9]:
#check datatypes
df.dtypes

Title              object
Genres             object
Rank               object
Popularity          int64
Score             float64
Episodes           object
Episode length     object
Release Date       object
dtype: object

In [10]:
df['Episodes'].isnull().values.any()

False

In [11]:
# convert to nuemeric
df['Episodes'] = pd.to_numeric(df['Episodes'])

ValueError: ignored

In [12]:
# The problem is the string "unknown", so remove all the rows that contain it
#df = df[df["Episodes"].str.contains("Unknown")== False] --> this gave a warning/error
df = df.loc[df['Episodes'] != "Unknown"]

In [13]:
# now convert to nuemeric
df['Episodes'] = pd.to_numeric(df['Episodes'])

#check datatypes
df.dtypes

Title              object
Genres             object
Rank               object
Popularity          int64
Score             float64
Episodes            int64
Episode length     object
Release Date       object
dtype: object

In [14]:
# objects are categorical variables
# we use select_dtypes()
df = df.select_dtypes(exclude=['object'])

In [15]:
# to drop rows with null values we use .dropna()
df = df.dropna()

In [17]:
df.to_csv(r'MALratings edited.csv', index=False)